src/Compiler/Microsoft.NET.Sdk.Razor.SourceGenerators/RazorSourceGenerator.cs

```diff
@@ -94,52 +92,22 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
-                {
-
-                    var (((generatedDeclarationSyntaxTree, allGeneratedDeclarationSyntaxTrees), compilation), razorSourceGeneratorOptions) = pair;
-                    RazorSourceGeneratorEventSource.Log.DiscoverTagHelpersFromComponentStart(generatedDeclarationSyntaxTree.FilePath);
-
-                    var tagHelperFeature = new StaticCompilationTagHelperFeature();
-                    var discoveryProjectEngine = GetDiscoveryProjectEngine(compilation.References.ToImmutableArray(), tagHelperFeature);
-
-                    var compilationWithDeclarations = compilation.AddSyntaxTrees(allGeneratedDeclarationSyntaxTrees);
-
-                    // try and find the specific root class this component is declaring, falling back to the assembly if for any reason the code is not in the shape we expect
-                    ISymbol targetSymbol = compilationWithDeclarations.Assembly;
-                    var root = generatedDeclarationSyntaxTree.GetRoot(ct);
-                    if (root is CompilationUnitSyntax { Members: [NamespaceDeclarationSyntax { Members: [ClassDeclarationSyntax classSyntax, ..] }, ..] })
-                    {
-                        var declaredClass = compilationWithDeclarations.GetSemanticModel(generatedDeclarationSyntaxTree).GetDeclaredSymbol(classSyntax, ct);
-                        Debug.Assert(declaredClass is null || declaredClass is { AllInterfaces: [{ Name: "IComponent" }, ..] });
-                        targetSymbol = declaredClass ?? targetSymbol;
-                    }
-
-                    tagHelperFeature.Compilation = compilationWithDeclarations;
-                    tagHelperFeature.TargetSymbol = targetSymbol;
-
-                    var result = tagHelperFeature.GetDescriptors();
-                    RazorSourceGeneratorEventSource.Log.DiscoverTagHelpersFromComponentStop(generatedDeclarationSyntaxTree.FilePath);
-                    return result;
-                });
-
             var tagHelpersFromCompilation = compilation
+                .Combine(generatedDeclarationSyntaxTrees.Collect())
                 .Combine(razorSourceGeneratorOptions)
                 .Select(static (pair, _) =>
                 {
                     RazorSourceGeneratorEventSource.Log.DiscoverTagHelpersFromCompilationStart();
 
-                    var (compilation, razorSourceGeneratorOptions) = pair;
+                    var ((compilation, generatedDeclarationSyntaxTrees), razorSourceGeneratorOptions) = pair;
 
                     var tagHelperFeature = new StaticCompilationTagHelperFeature();
                     var discoveryProjectEngine = GetDiscoveryProjectEngine(compilation.References.ToImmutableArray(), tagHelperFeature);
 
-                    tagHelperFeature.Compilation = compilation;
-                    tagHelperFeature.TargetSymbol = compilation.Assembly;
+                    var compilationWithDeclarations = compilation.AddSyntaxTrees(generatedDeclarationSyntaxTrees);
```

Comment:
This is a pretty significant performance de-opt. I don't think I'm understanding why it's necessary; what's actually causing the errors here?

Reply:
Right, forgot to explain, sorry. For example, consider the test added in this PR. The component is discovered both in `tagHelpersFromComponents` step and `tagHelpersFromCompilation` step. In the latter, the descriptor is different than in the former, because the component from the `.cs` file has a property `ChildContent` (appears as `BoundAttribute` on the descriptor), so both tag helper descriptors remain.

Yes, we could deduplicate them somehow, I guess. But I'm not sure if this is such de-opt, since `compilation.AddSyntaxTrees(generatedDeclarationSyntaxTrees);` was already present (in a different step).

Reply:
Not sure what you mean by previous code. But before the perf improvements, the code did a full discovery on the entire compilation every time a razor file changed:

https://github.com/dotnet/razor/blob/fcae93171c28d0a3935423be5584176e3ce29bbc/src/Compiler/Microsoft.NET.Sdk.Razor.SourceGenerators/RazorSourceGenerator.cs#L94-L131

Right now in `main`, you're right it's better, the discovery still happens for the full compilation (since fixing another regression in https://github.com/dotnet/razor/pull/8614) but it's scoped per component (`tagHelperFeature.TargetSymbol = targetSymbol;`):

https://github.com/dotnet/razor/blob/5803c80ec6cbc7e3b8640f8f9ea154c2eb0d98e1/src/Compiler/Microsoft.NET.Sdk.Razor.SourceGenerators/RazorSourceGenerator.cs#L97-L128